### DISCUSSION TITLE
Providing Models in SafeTensors format

### DISCUSSION THREAD

**Discussion by ohensf** [openai/whisper#1525] [Ideas]
Hi,

We are currently utilizing models in our project stored in pickle format. However, after exploring the advantages of SafeTensors in terms of improved security we believe that it will provide us with an extra layer of security and protection against potential vulnerabilities.
model link: [base](https://openaipublic.azureedge.net/main/whisper/models/ed3a0b6b1c0edf879ad9b11b1af5a0e6ab5db9205f891f668f8b0e6c6326e34e/base.pt), [base.en](https://openaipublic.azureedge.net/main/whisper/models/25a8566e1d0c1e2231d1c762132cd20e0f96a85d16145c3a00adf5d1ac670ead/base.en.pt)

We wish to inquire about the plans or timeline, if any, regarding the update of models from pickle to SafeTensors:
* Are there any plans to update the models to SafeTensors?
* If yes, is there a projected timeline for this migration?

**Comment by phineas-pta** [comment]
# u can do it yourself, not too much difficult

## simple code

convert to SafeTensors:
```python
import whisper
import safetensors.torch

model = whisper.load_model("large")
safetensors.torch.save_model(model, "large.safetensors")
```

load from SafeTensors:
```python
import whisper
import safetensors.torch

model = whisper.model.Whisper(whisper.model.ModelDimensions(
    n_mels        = 80,
    n_vocab       = 51865,
    n_audio_ctx   = 1500,
    n_audio_state = 1280,
    n_audio_head  = 20,
    n_audio_layer = 32,
    n_text_ctx    = 448,
    n_text_state  = 1280,
    n_text_head   = 20,
    n_text_layer  = 32
))
# model.to("cuda")

safetensors.torch.load_model(model, "large.safetensors")
model.set_alignment_heads(b"ABzY8zd+h!0{>%R7=D0pU<_bnWW*tkYAhobTNnu$jnkEkXqp)j;w1Tzk)UH3X%SZd&fFZ2fC2yj")

# model.transcribe(...)
```

## models info

value for alignment heads can be found in https://github.com/openai/whisper/blob/main/whisper/__init__.py
```python
_ALIGNMENT_HEADS = {
    "tiny.en"  : b"ABzY8J1N>@0{>%R00Bk>$p{7v037`oCl~+#00",
    "tiny"     : b"ABzY8bu8Lr0{>%RKn9Fp%m@SkK7Kt=7ytkO",
    "base.en"  : b"ABzY8;40c<0{>%RzzG;p*o+Vo09|#PsxSZm00",
    "base"     : b"ABzY8KQ!870{>%RzyTQH3`Q^yNP!>##QT-<FaQ7m",
    "small.en" : b"ABzY8>?_)10{>%RpeA61k&I|OI3I$65C{;;pbCHh0B{qLQ;+}v00",
    "small"    : b"ABzY8DmU6=0{>%Rpa?J`kvJ6qF(V^F86#Xh7JUGMK}P<N0000",
    "medium.en": b"ABzY8usPae0{>%R7<zz_OvQ{)4kMa0BMw6u5rT}kRKX;$NfYBv00*Hl@qhsU00",
    "medium"   : b"ABzY8B0Jh+0{>%R7}kK1fFL7w6%<-Pf*t^=N)Qr&0RR9",
    "large-v1" : b"ABzY8r9j$a0{>%R7#4sLmoOs{s)o3~84-RPdcFk!JR<kSfC2yj",
    "large-v2" : b"ABzY8zd+h!0{>%R7=D0pU<_bnWW*tkYAhobTNnu$jnkEkXqp)j;w1Tzk)UH3X%SZd&fFZ2fC2yj",
    "large"    : b"ABzY8zd+h!0{>%R7=D0pU<_bnWW*tkYAhobTNnu$jnkEkXqp)j;w1Tzk)UH3X%SZd&fFZ2fC2yj",
}
```

value for dimensions can be found using `torch.load("PATH to checkpoint like ~/.cache/whisper/large-v2.pt")["dims"]`

here the values:
<table>
	<thead><tr>
		<th>&nbsp;</th>
		<th>tiny</th>
		<th>base</th>
		<th>small</th>
		<th>medium</th>
		<th>large</th>
	</tr></thead>
	<tbody>
		<tr>
			<td><code>n_mels</code></td>
			<td colspan="5" align="center">80</td>
		</tr>
		<tr>
			<td><code>n_vocab</code></td>
			<td colspan="5" align="center">51865 if multilingual models<br />51864 if english-only models</td>
		</tr>
		<tr>
			<td><code>n_audio_ctx</code></td>
			<td colspan="5" align="center">1500</td>
		</tr>
		<tr>
			<td><code>n_audio_state</code></td>
			<td>384</td>
			<td>512</td>
			<td>768</td>
			<td>1024</td>
			<td>1280</td>
		</tr>
		<tr>
			<td><code>n_audio_head</code></td>
			<td>6</td>
			<td>8</td>
			<td>12</td>
			<td>16</td>
			<td>20</td>
		</tr>
		<tr>
			<td><code>n_audio_layer</code></td>
			<td>4</td>
			<td>6</td>
			<td>12</td>
			<td>24</td>
			<td>32</td>
		</tr>
		<tr>
			<td><code>n_text_ctx</code></td>
			<td colspan="5" align="center">448</td>
		</tr>
		<tr>
			<td><code>n_text_state</code></td>
			<td colspan="5" align="center">same as <code>n_audio_state</code></td>
		</tr>
		<tr>
			<td><code>n_text_head</code></td>
			<td colspan="5" align="center">same as <code>n_audio_head</code></td>
		</tr>
		<tr>
			<td><code>n_text_layer</code></td>
			<td colspan="5" align="center">same as <code>n_audio_layer</code></td>
		</tr>
	</tbody>
</table>

**Reply by ohensf** [reply]
Great, it is really helpful. 🙏  thanks!

**Reply by ohensf** [reply]
@phineas-pta One thing that we notice is that when running the pickle version vs. the SafeTensors version, we are sometimes getting different results (one word, sometimes a few...). 
Do you know if the SafeTensors conversion should impact the quality of the results? Or is it just a matter of configuration?

**Reply by phineas-pta** [reply]
#81

**Reply by srezasm** [reply]
How should I set the `whisper.model.ModelDimensions` parameters for small model?

**Comment by phineas-pta** [comment]
huggingface also created safetensor version of whisper, but to be used with their pipeline interface
